File: src/main/java/nz/co/searchwellington/model/Resource.java
package nz.co.searchwellington.model;

import java.util.Date;
import java.util.Set;

public interface Resource {
	
    public int getId();
    public void setId(int id);
    
    public String getName();
    public void setName(String name);
    
    public String getUrl();
    public void setUrl(String url);
  
    public String getDescription();
        
    public String getType();
    
    public Date getDate();
    public void setDate(Date date);
    
    public void setDescription(String description);
    
    public int getHttpStatus();
    public void setHttpStatus(int httpStatus);
    public Date getLastScanned();
    public void setLastScanned(Date lastScanned);
    
    public Date getLastChanged();
    public void setLastChanged(Date lastChanged);
    
    public Set<DiscoveredFeed> getDiscoveredFeeds();
    public void setDiscoveredFeeds(Set<DiscoveredFeed> discoveredFeeds);
    
    public Date getLiveTime();
    public void setLiveTime(Date time);
    
	public Date getEmbargoedUntil();
	public void setEmbargoedUntil(Date embargoedUntil);
    
	public Geocode getGeocode();
	public void setGeocode(Geocode geocode);
	public String getUrlWords();
	public void setUrlWords(String urlWords);
	
	public User getOwner();
	public void setOwner(User owner);

	public boolean isHeld();
	public void setHeld(boolean held);
	
}

File: src/main/scala/nz/co/searchwellington/model/Resource.scala
package nz.co.searchwellington.model

import nz.co.searchwellington.model.geo.Geocode

import java.util.Date
import reactivemongo.api.bson.BSONObjectID

trait Resource extends Tagged {

  val _id: BSONObjectID
  val id: String
  val `type`: String
  var title: String
  var description: Option[String]
  val page: String
  var httpStatus: Option[HttpStatus]
  var date: Option[Date]  // TODO backfill ~ 100 records with no date and make mandatory
  var last_scanned: Option[Date]
  var last_changed: Option[Date]
  var live_time: Option[Date]
  var embargoed_until: Option[Date]
  var held: Boolean
  var url_words: Option[String]
  var geocode: Option[Geocode]
  var owner: Option[BSONObjectID]
  val resource_tags: Seq[Tagging]

  def setHttpStatus(status: Int, redirecting: Boolean): Unit = {
    if (httpStatus.nonEmpty) {
      httpStatus = httpStatus.map(_.copy(status = status, redirecting = redirecting))
    } else {
      httpStatus = Some(HttpStatus(status, redirecting = redirecting))
    }
  }

  def setLastScanned(lastScanned: Date): Unit = this.last_scanned = Some(lastScanned)
  def setLastChanged(lastChanged: Date): Unit = this.last_changed = Some(lastChanged)

}
